test/format/snap.c

```diff
@@ -492,3 +649,205 @@ snap_repeat_single(WT_CURSOR *cursor, TINFO *tinfo)
+     * that may be relevant.
+     *
+     * We need to limit how far back we can examine. For example, if we see a modification for key X
+     * in thread T, there may in fact have been a more recent modification for key X in thread U,
+     * but we can't see it because the snap list for U has wrapped past the modification for key X.
+     * We need to look at the oldest timestamp recorded in the snaps for each thread, and use the
+     * maximum of all of these for our limit. That's the newest, oldest timestamp.
+     */
+    newest_oldest_ts = 0ULL;
+    sorted_end = sorted_snaps = dcalloc(SNAP_LIST_SIZE * tinfo_count, sizeof(SNAP_OPS *));
+
+    for (i = 0, tinfop = tinfo_array; i < tinfo_count; ++i, ++tinfop) {
+        tinfo = *tinfop;
+        /*
+         * If this thread has knowledge of the current stable timestamp, that means its "other" snap
+         * list stores up to the stable timestamp, it's the one we want to use. If this thread
+         * doesn't yet have knowledge of the current stable, that means the current snap list is the
+         * one we want.
+         */
+        if (tinfo->stable_ts != g.stable_timestamp)
+            state = tinfo->s;
+        else if (tinfo->s == &tinfo->snap_states[0])
+            state = &tinfo->snap_states[1];
+        else
+            state = &tinfo->snap_states[0];
+        oldest_ts = UINT64_MAX;
+        for (snap = state->snap_state_list; snap < state->snap_state_end; ++snap) {
+            /*
+             * Only keep entries that aren't cleared out and may have relevant timestamps. We don't
+             * fully know which timestamps are relevant, since we haven't computed the newest oldest
+             * yet. We do keep entries that are not marked repeatable, we won't retry unrepeatable
+             * reads, but we need them to invalidate keys that we shouldn't check.
+             */
+            if (snap->op != 0 && snap->ts != 0 && snap->ts <= g.stable_timestamp) {
+                oldest_ts = WT_MIN(oldest_ts, snap->ts);
+                *sorted_end++ = snap;
+            }
+        }
+
+        /*
+         * If there aren't any entries older than the stable timestamp, we've wrapped around. This
+         * thread may have made changes to any key right up to the stable time that have now been
+         * overwritten. There's no way to get an accurate accounting, so we skip checking for this
+         * run.
+         */
+        if (oldest_ts == UINT64_MAX)
+            goto cleanup;
+        newest_oldest_ts = WT_MAX(newest_oldest_ts, oldest_ts);
+    }
+
+    __wt_qsort(
+      sorted_snaps, (size_t)(sorted_end - sorted_snaps), sizeof(SNAP_OPS *), compare_snap_ts);
+
+    /*
+     * Start a new transaction. Verify all repeatable records. Discard the transaction.
+     */
+    while ((ret = session->begin_transaction(session, "isolation=snapshot")) == WT_CACHE_FULL)
+        __wt_yield();
+    testutil_check(ret);
```

Comment:
`wiredtiger_begin_transaction` in `format.i` does this for you.

Reply:
Fixed.